Experimental namespace processing in java

ghuron opened the pull request:
Right now tool cannot work due to change in db replicas (see https://phabricator.wikimedia.org/T188506 for details). In order to get it back to operation, I've made 3 modifications:
1. Comment pieces, related to __extra databases (probably will need additional cleanup here)
2. Modify .ttl so it operates with "fake" namespaces like (e.g. /wiki/ns14:) instead of "real" (/wiki/Category:)
3. Implement java converter from "real" to "fake" namespaces and back based on mediawiki api calls (see comment from Stas Malyshev in phabricator above)

I've tested it locally running simple queries like `select * {<https://en.wikipedia.org/wiki/Category:Burials_by_place> ?predicate ?object}` Running a bit slow, but we can work on perfornance once tool is accessible from WQS via federation.

ghuron committed:
Create MWNamespace.java

ghuron committed:
Update mapping.ttl

ghuron committed:
Update SPARQLActions.java

ghuron committed:
Update RepositoryFactory.java

Tpt commented on src/main/java/org/mediawiki/sparql/mwontop/http/SPARQLActions.java:
```diff
@@ -134,7 +134,26 @@ private Response evaluateTupleQuery(TupleQuery query, Request request) {
+                            MapBindingSet newSet = new MapBindingSet();
+                            for (Binding binding : oldSet) {
+                                Value value = binding.getValue();
+                                if (binding.getValue() instanceof SimpleIRI) {
```
It's probably safer to use the `org.eclipse.rdf4j.model.IRI` interface here (if Ontop or something else switches to an other IRI implementation).

Tpt commented on src/main/java/org/mediawiki/sparql/mwontop/http/SPARQLActions.java:
```diff
@@ -134,7 +134,26 @@ private Response evaluateTupleQuery(TupleQuery query, Request request) {
+                        TupleQueryResultWriter writer = format.getService().getWriter(outputStream);
+                        TupleQueryResult result = query.evaluate();
+
+                        writer.startQueryResult(result.getBindingNames());
```
Maybe move the conversion logic in its own function that takes and return a BindingSet?

Tpt commented on src/main/java/org/mediawiki/sparql/mwontop/http/MWNamespace.java:
```diff
@@ -0,0 +1,73 @@
+
+    private static Pattern ns_regex = Pattern.compile("//([^/]*)/wiki/([^:]*:)?");
+
+    static String transformNamespace(String input) {
```
Maybe `encodeNamespace` to make the direction clear?

Tpt reviewed (commented):
Thank you very much for the change. 
What you be also nice to encode/decode main namespace URIs like `http://en.wikipedia.org/wiki/Foo` as `http://en.wikipedia.org/wiki/ns0:Foo` in order to remove all the special cases for the main namespace in  `mapping.ttl`. It should also make the generated queries smaller and faster because currently page titles like "ns4:Foo" are considered as also maybe taking part of the main namespace and so both the SQL queries for the main namespace and for the other namespaces are used.

ghuron committed:
Refactoring based on feedback

Extracted all logic for query execution, results transformation and writing to output stream into separate method

ghuron committed:
Clarifies purpose of public method

Change "transform" to more neutral "mutate" and write description in javadoc

ghuron commented:
@Tpt I think your suggestion to streamline .ttl with /ns0: notation makes a lot of sense, but I'm a little bit afraid of doing such radical changes without first seeing that mw2sparql returned back to operation. I cannot test this patch with federated queries, and if it will there will be issues in production, I'd prefer to have less changes to debug, not more.
Can we postpone that for a few days? I promise I will implement that (as well as other performance changes - what I see under "DEBUG i.u.i.o.a.r.impl.QuestQueryProcessor - Resulting native query:" doesn't look efficient for me)

Tpt commented:
The code is deployed on Tools Labs and seems to work. Thank you!
